driver-core/src/main/com/mongodb/internal/connection/SocketStreamHelper.java

```diff
@@ -78,19 +107,21 @@ static void initialize(final Socket socket, final InetSocketAddress inetSocketAd
         socket.connect(inetSocketAddress, settings.getConnectTimeout(MILLISECONDS));
     }
 
-    @SuppressWarnings("unchecked")
-    private static void setExtendedSocketOptions(final Socket socket) {
-        if (Arrays.stream(ExtendedSocketOptions.class.getDeclaredFields()).anyMatch(f -> f.getName().equals(TCP_KEEPCOUNT))) {
-            try {
-                Method setOptionMethod = Socket.class.getMethod("setOption", SocketOption.class, Object.class);
-                setOptionMethod.invoke(socket, ExtendedSocketOptions.class.getDeclaredField(TCP_KEEPCOUNT).get(null),
-                        TCP_KEEPCOUNT_LIMIT);
-                setOptionMethod.invoke(socket, ExtendedSocketOptions.class.getDeclaredField(TCP_KEEPIDLE).get(null),
-                        TCP_KEEPIDLE_DURATION);
-                setOptionMethod.invoke(socket, ExtendedSocketOptions.class.getDeclaredField(TCP_KEEPINTERVAL).get(null),
-                        TCP_KEEPINTERVAL_DURATION);
-            } catch (Throwable t) {
+    static void setExtendedSocketOptions(final Socket socket) {
+        try {
+            if (SET_OPTION_METHOD != null) {
+                if (KEEP_COUNT_OPTION != null) {
+                    SET_OPTION_METHOD.invoke(socket, KEEP_COUNT_OPTION, TCP_KEEPCOUNT_LIMIT);
```

Comment:
This PR made me think, why does the driver enable the TCP keep alive mechanism when connecting to a MongoDB server? [RFC 1122. 4.2.3.6 TCP Keep-Alives](https://www.rfc-editor.org/rfc/rfc1122#page-101):

> A TCP keep-alive mechanism should only be invoked in
> server applications that might otherwise hang
> indefinitely and consume resources unnecessarily if a
> client crashes or aborts a connection during a network
> failure.

Reply:
There is discussion of the benefits in https://jira.mongodb.org/browse/DRIVERS-383